### ISSUE TITLE
Flag vs. access key id/secret to make presign url work...

### ISSUE THREAD

**Issue by asautins** [localstack/localstack#3544]
[ ] bug report
[X] feature request

# Detailed description

The documentation states:

NOTE: Please use test as Access key id and secret Access Key to make S3 presign url work. We have added presign url signature verification algorithm to validate the presign url and its expiration. 

If I want pre-signed URLs enforce across localstack I need to ensure that everyone logging into localstack is using the correct credentials.  What would be easier to enforce would be to have a flag ( e.g., S3_PRESIGN_URL=1 ) vs. setting the access key and secret.
...

## Expected behavior

If a flag is set ( e.g., S3_PRESIGN_URL = 1) pre-signed URLs are used no matter what the access key id/secret are set to.
...

## Actual behavior

Currently pre-signed URLs require test/test as the access key id and secret.
...

# Steps to reproduce

## Command used to start LocalStack

...

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

...

**Comment by asautins** [comment]
This issue was noticed when [DoesS3BucketExistAsync](https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/S3/MS3UtilDoesS3BucketExistAsyncIS3String.html) started to fail silently ( returning true when the bucket does not exist ).  Reading the documentation:

> Determines whether an S3 bucket exists or not. This is done by: 1. Creating a PreSigned Url for the bucket. To work with > Signature V4 only regions, as well as Signature V4-optional regions, we keep the expiry to within the maximum for V4 (which is > one week). 2. Making a HEAD request to the Url

The documentation above indicates that the above method requires pre-signed url.  It would be good to have a global flag to enable pre-signing of URLs to not infer from the id/secret if certain methods would work correctly.

**Comment by asautins** [comment]
I wanted to just post a follow up after having thought about it for a little while.

It's interesting, the documentation says to use test/test for username and password.  From the documentation at https://github.com/localstack/localstack it says:

> NOTE: Please use test as Access key id and secret Access Key to make S3 presign url work. We have added presign url signature verification algorithm to validate the presign url and its expiration.  You can configure credentials into the system environment using export command in the linux/Mac system. You also can add credentials in ~/.aws/credentials file directly.

That would seem to indicate that localstack is expecting one credential ( not different users accessing S3 concurrently ) and that user needs to be test.

If the user needs to be test it would be great if there were quick way of letting the user know they are not using test.  It does appear that accessing some functionality on S3 as a user other than test fails silently ( e.g. the method DoesS3BucketExistAsync mentioned above ).  The failing silently is concerning in that it's not clear where to go to understand that to resolve it the id/pass should be set to test/test.

Ideally I would like to be able to any credentials to access S3.  If there is a limitation that credentials need to be test/test it would be nice to error out in a way other than silently failing.  Something to point the user to how to correct the issue.

My suggestion in this ticket originally was a flag that just said don't validate pre-signed URLS.  From looking at the code it appears that s3_listener.py contains the method authenticate_presign_url that calculates the signature based off of the test/test credentials.  What would be the issue with a flag that, when set, caused authenticate_presign_url to always return true or even have the TEST_AWS_ACCESS_KEY_ID and TEST_AWS_ACCESS_KEY_ID be configurable in the environment.

**Comment by whummer** [comment]
Thanks for reporting @asautins, and apologies for the delay . We're currently looking into some issues related to S3 presigned URLs in more detail.

Generally, the assumption is that access key and secret key are both set to the value `test` for your client which generates the presigned URL. We're currently hardening the logic, to make sure the signature calculation is stable and matches for different host names, header values, and other configuration options.

Meanwhile, as a workaround we have introduced a **new configuration value `S3_SKIP_SIGNATURE_VALIDATION=1`** which you can enable in your LocalStack container environment, to temporarily skip the signature validation.

Hope that helps - please keep us posted how it goes. Thanks!

**Comment by asautins** [comment]
@whummer,

Thank you for the update.  

One question, is it true that if I need to set the secret key and access key both to `test` that I can only access S3 as one user.  I would suppose that would mean testing any security or access control through localstack wouldn't be possible if that is the case.  Is that true?

I appreciate the workaround.  We have already changed our secret key and access keys to `test` so while it's a great addition, we do not currently have a need for it.  It would, however, have helped us quite a bit if it were there and I appreciate the addition.

One additional note is that in our scenario, the call to DoesS3BucketExistAsync would fail silently.  After deeper reading of the documentation we realized that we should be using the secret key and access key of `test`.  I wonder if it would be helpful to log a warning or error if test is expected/required but not being provided.

Thank you.
